### PR TITLE
tutorial for porting ethereum dapps to avalanche

### PR DESCRIPTION
**Pull request by jpop32**

### CODE REVIEW

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+## Platform basics
+
+Avalanche is a [network of networks](../../../learn/platform-overview/README.md). It means that it is not a single chain running a single, uniform type of blocks. It contains multiple subnets, each running one of more heterogenous chains. But, to run an Ethereum dapp on a low-fee, fast network with instant finality, we don't need to concern ourselves with that right now. Using the link above you can find out more if you wish, but all you need to know right now is that one of the chains running on Avalanche Primary Network is the C-Chain (contract chain).
+
+C-Chain runs a fork of [go-ethereum](https://geth.ethereum.org/docs/rpc/server)\) called [coreth](https://github.com/ava-labs/coreth) that has the networking and consensus portions replaced with Avalanche equivalents. What's left is the Solidity VM, which runs the smart contracts and manages data structures and blocks on the chain. As a result you get a blockchain that can run all the Solidity smart contracts from Ethereum, but with much greater transaction bandwidth and instant finality that [Avalanche's revolutionary consensus](../../../learn/platform-overview/avalanche-consensus.md) enables.
```
[the remark this answers is not in the document]
Comment (dasconnor) [reply build/tutorials/platform/port-your-ethereum-dapp.md:11]: technically, I believe we support Vyper but not one has tried this

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+* **Symbol**: `AVAX`
+* **Explorer**: [https://cchain.explorer.avax-test.network](https://cchain.explorer.avax-test.network/)
+
+In your application's web interface, you can [add Avalanche programmatically](../smart-contracts/add-avalanche-to-metamask-programmatically.md) so your users don't have to enter the network data manually. To see the adding custom network flow in action, check out [Pangolin DEX](https://app.pangolin.exchange/).
+
+### Using the public API nodes
+
+Instead of proxying network operations through MetaMask, you can use the public API, which consists of a number of load-balanced AvalancheGo nodes behind a load balancer.
+
+The C-Chain API endpoint is [https://api.avax.network/ext/bc/C/rpc](https://api.avax.network/ext/bc/C/rpc) for the main net and [https://api.avax-test.network/ext/bc/C/rpc](https://api.avax-test.network/ext/bc/C/rpc) for the testnet.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:49]: nit: `the main net` -> `mainnet`

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+### Running your own node
+
+If you don't want for your dapp to depend on a centralized service you don't control, you can run your own node and access the network that way.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:53]: I'd also add that running your own node is an easy way to get around rate-limiting

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+Being an Ethereum-compatible blockchain, all of the usual Ethereum developer tools and environments can be used to develop and deploy dapps for Avalanche's C-Chain. 
+
+### Remix
+
+[Here](../smart-contracts/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask.md) is a tutorial for using Remix to deploy smart contracts on Avalanche. It relies on MetaMask for access to Avalanche network.
+
+### Truffle
+
+You can also use Truffle to test and deploy smart contracts on Avalanche. Find out how in [this](../smart-contracts/using-truffle-with-the-avalanche-c-chain.md) tutorial.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:73]: Truffle support might be broken right now, but I'm not sure

Comment (dasconnor) [reply build/tutorials/platform/port-your-ethereum-dapp.md:73]: not sure there's an action required here, just mentioning

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+### Hardhat
+
+Hardhat is the newest development and testing environment for Solidity smart contracts, and the one our developers use the most. Due to its superb testing support, it is the recommended way of developing for Avalanche.
+
+[Here](https://github.com/ava-labs/avalanche-smart-contract-quickstart) is a quickstart repository that our developers use to start new projects.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:79]: mention that it is configured for avalanche so no additional setup is required

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+## Avalanche Explorer
+
+Essential part of the smart contract development environment is the explorer, that indexes and serves blockchain data. Main net C-Chain explorer is available at [https://cchain.explorer.avax.network/](https://cchain.explorer.avax.network/) and test net explorer at [https://cchain.explorer.avax-test.network](https://cchain.explorer.avax-test.network/). Besides the web interface, it also exposes the standard [Ethereum JSON RPC API](https://eth.wiki/json-rpc/API).
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:83]: Mainnet again

Comment (dasconnor) [reply build/tutorials/platform/port-your-ethereum-dapp.md:83]: testnet too

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+## Avalanche Faucet
+
+For development purposes, you will need test tokens. Avalanche has a [Faucet](https://faucet.avax-test.network/) that drips tet tokens to address of your choice. Paste your C-Chain address there.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:87]: tet?

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+## Gotchas and things to look out for
+
+Avalanche Platform's C-Chain is EVM-compatible, but it is not identical. There are some differences you need to be aware of, otherwise you may create subtle bugs or inconsistencies in how you dapps behave.
+
+Here are the main differences you should be aware of.
+
+### Measuring time
+
+It is customary on Ethereum to use block height progress as a proxy for time. You should not do that on Avalanche. Chains on Avalanche are quiescent, meaning that if there is no activity, there are no blocks produced. The opposite is also true, if there is great amount of activity, blocks are produced very fast. Because of that, you should not measure the passage of time by the amount of blocks that are produced. You both get inaccurate results, and be open to manipulation by third parties.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:109]: The last sentence in this paragraph isn't grammatically correct. I'd recommend: The results will not be accurate, and your contract may be manipulated by third parties.

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+Instead of block rate, you should measure time simply by reading the timestamp attribute of the produced blocks. Timestamps are guaranteed to be monotonically increasing, and to be within 30s of the real time.
+
+### Finality
+
+On Ethereum, blockchain can be reorganized and blocks can be orphaned, so you cannot rely on the fact that a block has been accepted until it is several blocks further from the tip (usually, it is presumed that blocks 6 places deep are safe). That is not the case on Avalanche. Blocks are either accepted or rejected within a second or two. And once the block has been accepted, it is final, and cannot be replaced, dropped or modified. So the concept of 'number of confirmations' on Avalanche is not used. As soon as a block is accepted and available in the explorer, it is final.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:115]: insert `the` as the third word

File: build/tutorials/platform/port-your-ethereum-dapp.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+Instead of block rate, you should measure time simply by reading the timestamp attribute of the produced blocks. Timestamps are guaranteed to be monotonically increasing, and to be within 30s of the real time.
+
+### Finality
+
+On Ethereum, blockchain can be reorganized and blocks can be orphaned, so you cannot rely on the fact that a block has been accepted until it is several blocks further from the tip (usually, it is presumed that blocks 6 places deep are safe). That is not the case on Avalanche. Blocks are either accepted or rejected within a second or two. And once the block has been accepted, it is final, and cannot be replaced, dropped or modified. So the concept of 'number of confirmations' on Avalanche is not used. As soon as a block is accepted and available in the explorer, it is final. 
+
+### Gas price
+
+Gas on Avalanche is burned. Validators don't keep the gas for themselves (they get reward for staking), so the dynamics of 'gas wars' where higher priced transactions are included first is non-existent. Therefore, there is never a need to put a higher gas price on your transactions. You'll only be burning gas in vain.
```
Comment (dasconnor) [inline build/tutorials/platform/port-your-ethereum-dapp.md:119]: reward -> rewarded

File: build/tutorials/platform/port-your-ethereum-dapp.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpop32) [reply build/tutorials/platform/port-your-ethereum-dapp.md:11]: reworded. wouldn't mention vyper until we're sure.

**Review by dasconnor — APPROVED** [review]
lgtm

**Review by cgcardona — APPROVED** [review]

**Pull request merged**